query/src/main/scala/filodb/query/exec/PromQlExec.scala

```diff
@@ -0,0 +1,117 @@
+    * implementation of the operation represented by this exec plan
+    * node
+    */
+  override protected def doExecute(source: ChunkSource, dataset: Dataset, queryConfig: QueryConfig)
+                                  (implicit sched: Scheduler, timeout: FiniteDuration): Observable[RangeVector] = ???
+
+  /**
+    * Sub classes should implement this with schema of RangeVectors returned
+    * from doExecute() abstract method.
+    */
+  override protected def schemaOfDoExecute(dataset: Dataset): ResultSchema = ???
+
+  val columns: Seq[ColumnInfo] = Seq(ColumnInfo("timestamp", ColumnType.LongColumn),
+    ColumnInfo("value", ColumnType.DoubleColumn))
+  val recSchema = SerializableRangeVector.toSchema(columns)
+  val builder = SerializableRangeVector.toBuilder(recSchema)
+
+  override def execute(source: ChunkSource,
+                       dataset: Dataset,
+                       queryConfig: QueryConfig)
+                      (implicit sched: Scheduler,
+                       timeout: FiniteDuration): Task[QueryResponse] = {
+
+    val queryResponse = PromQlExec.httpClient(params).map(response => {
+
+      response.unsafeBody match {
+        case Left(error) => QueryError(id, new IllegalStateException(error.message))
+        case Right(response) => toQueryResponse(response.data, id)
+      }
+
+    })
+    Task.fromFuture(queryResponse)
+  }
+
+  def toQueryResponse(data: Data, id: String): QueryResponse = {
+
+    val schema = ResultSchema(columns, 1)
+    val rangeVectors = data.result.map {
+      r => {
+        val rv = new RangeVector {
+          override def key: RangeVectorKey = CustomRangeVectorKey(r.metric.map {
+            m => {
+              m._1.utf8 -> m._2.utf8
+            }
+          })
+
+          override def rows: Iterator[RowReader] = r.values.map { v => {
+            new TransientRow(v.timestamp * 1000, v.value)
+          }
+          }.iterator
+        }
+        SerializableRangeVector(rv, builder, recSchema, printTree(false))
+      }
+    }
+    QueryResult(id, schema, rangeVectors)
+  }
+}
+
+object PromQlExec extends  StrictLogging{
+
+  import com.softwaremill.sttp._
+  import io.circe.generic.auto._
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+  // DO NOT REMOVE PromCirceSupport import below assuming it is unused - Intellij removes it in auto-imports :( .
+  // Needed to override Sampl case class Encoder.
+  import PromCirceSupport._
+  implicit val backend = AkkaHttpBackend()
+
+  def httpClient(params: PromQlQueryParams):
+  Future[Response[scala.Either[DeserializationError[io.circe.Error], SuccessResponse]]] = {
+    var urlParams = Map("query" -> params.promQl, "start" -> params.start, "end" -> params.end, "step" -> params.step,
+      "processFailure" -> params.processFailure)
+    if (params.spread.isDefined)
+      urlParams = urlParams + ("spread" -> params.spread.get)
+
+    val endpoint = params.promEndPoint
+    val url = uri"$endpoint?$urlParams"
```

Comment:
Do not do this way. attach query params via API. It will do url encoding for all the params.